### PR TITLE
Add Storage dashboard

### LINKED ISSUE #518: Improve quota display by distinguishing active and deleted but not purged space
More often than not users delete datasets or histories without purging resulting in unexpected quota being used.
In the case of full histories it's somewhat hidden how you can list which deleted but not purged entries you may have.

In addition, the set_user_disk_usage.py script only displays:

```
user <email> old usage: 1.5GB change: none
```

So it's also not easy for an administrator to see where the quota is being used.

I therefore suggest 2 enhancements, one on the set_user_disk_usage.py script that instead shows:

```
user <email> old usage: 1.5GB (active: 500MB, deleted: 1GB) change: none
```

Additionally the galaxy user interface could display a tooltip (clicking or hovering the quota widget) which would inform the user that a given % of their quota is being taken by unpurged datasets or histories.

from @Unode on [trello](https://trello.com/c/aILtCpdi/2212-improve-quota-display-by-distinguishing-active-and-deleted-but-not-purged-space)

### PR DESCRIPTION
**Pull request by davelopez**
This PR adds a new `Storage Dashboard` feature to the Galaxy client. It is my first "big" PR for the frontend so I will appreciate comments on style, best practices, patterns, and recommendations!

## General idea
The general idea is to have a **centralized place where the user can have an overview of the disk usage and quota and provide them with tools to easily recover space or manage their data**.

The current interface it's heavily inspired in [Google's storage management](https://one.google.com/storage) (maybe too heavily :sweat_smile:) by keeping it simple and guiding the user as much as possible through the options.

StorageDashboard

## Extensible
The Storage Manager view has been designed to be easily extended with new operations to discover and free space. Adding more cleanup operations is a matter of editing the `client/src/components/User/DiskUsage/Management/Cleanup/categories.js` and adding a new category or a new operation to an existing category like this:
```js
{
    name: _l("Old Items"), // Category name
    operations: [
        {
            id: "old_datasets", // Unique identifier for this operation
            name: _l("Old datasets"), // The display name
            description: _l("Datasets that haven't been deleted but were not modified in quite some time"),
            fetchSummary: async () => { // Callback that gets the total recoverable space and the number of items to clear
                return { totalSize: 0, totalItems: 0 };
            },
            fetchItems: async () => [], // Callback that fetches the actual items to clear
            cleanupItems: async () => {}, // Callback to purge/delete/clear a collection of items
        },
        {
            id: "old_histories",
            name: _l("Old histories"),
            description: _l("Histories that haven't been deleted but were not modified in quite some time"),
            fetchSummary: async () => {
                return { totalSize: 0, totalItems: 0 };
            },
            fetchItems: async () => [],
            cleanupItems: async () => {},
        },
    ],
},
```
This will add a new category `Old Items` with two operations to delete old datasets and histories. You just need to define proper functions for the `fetchSummary`, `fetchItems`, and `cleanupItems` callbacks.

Screenshot from 2022-01-04 14-24-54

## Open questions
- [x] **How this should work with anonymous users?** Right now this feature is not available to anonymous users because some of the API calls for retrieving datasets require a logged-in user. This probably can be changed but it's worth discussing.
   - From the discussion so far this feature makes sense only for registered users.
- [x] **What happens if the server has `allow_user_dataset_purge` disabled?** Well, right now it will throw an error when trying to permanently delete datasets. In this scenario, the "Deleted items" category may not be useful and could be conditionally hidden, we can also change the other non-deleted cleanup operations to `delete` instead of `purge` but then the quota change will not be reflected... some more material for discussion here :smile: 
   - Seems that `allow_user_dataset_purge` can be assumed to be always `true` so no issue here.

## Future work
- This PR only adds one cleanup operation, more can/will be added in different PRs as extensions.
- Add full support for multi quotas and different ObjectStores after https://github.com/galaxyproject/galaxy/pull/10977 is completed.
- In addition to the user-friendly usage cleanup utilities, we can add more options here to visually explore storage data with charts and other visualizations for power users.
- Provide more context information about a particular item when reviewing it. Maybe a popover with more details, or links to history, etc.
- Your ideas here...

## TODO (before is ready to review)
- [x] <del>Move backend changes to their own PRs to simplify review.</del> Improve and fix backend changes in this PR as Marius suggested.
- [x] Make it a bit more future-proof for https://github.com/galaxyproject/galaxy/pull/10977.
   - See https://github.com/galaxyproject/galaxy/pull/13113#issuecomment-1007598106
- [x] Add automated tests

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Delete some datasets
  - Click on the quota meter in the masthead or directly go to `http://127.0.0.1:8081/storage`

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

Fixes https://github.com/galaxyproject/galaxy/issues/518

### CODE REVIEW

**Comment by mvdbeek** [comment]
That is absolutely amazing and will be so useful when bringing in scratch storage, user-based storage, other kinds of dataset associations etc (#13110)!

> * How this should work with anonymous users?

IMO we can disable this, I don't think there's much of a point for anons to manage their storage.

> What happens if the server has allow_user_dataset_purge disabled? .... * the "Deleted items" category may not be useful and could be conditionally hidden

I think that's a good idea as the first step. My gut feeling is that (and it would be good to have feedback from admins on this point (ping @natefoo, @gmauro, @cat-bro, @hexylena )) the quota value should reflect the amount of non-deleted items (as opposed to non-purged items) if an admin turns off `allow_user_dataset_purge`. If an admin decides to not allow user dataset purge it should be their responsibility to clean up the data periodically. Users could game the storage during the cleanup period (i.e just delete + undelete in regular intervals if the cleanup script is set to cleanup after X days), but I doubt users would go to such extents. And it looks like user dataset purge is allowed on all of the  the usegalaxy.* instances, so this likely affects local instances that probably have more trust in their users / where it is not such a big deal if you have to manually warn a user that does this.

Again, super amazing stuff @davelopez!

**Comment by nsoranzo** [comment]
> What happens if the server has `allow_user_dataset_purge` disabled?

IMO we can drop `allow_user_dataset_purge`. It was added in 2011 in commit 1f5e8ffeefea7dea4a6e06a82e9907614f11a0c1 by @natefoo shortly after adding the possibility for users to purge their datasets. The options defaults to True since 2016 (commit 3d688e6bcf49e15fc12f45590250d34cccd0d8cb ), and I am not sure there is (still) a use case for preventing users from doing it.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/datasets.py]: I think that as it is you can only delete `HistoryDatasetAssociation` items. I would add the `src` parameter (could be `hda`, `ldda`, for `DatasetInstance` subclasses -- plus `dda` soon) here to control which model you can delete.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/datasets.py]: I am worried about users "destroying" their dataset collections by deleting members of their dataset collections here. That is a hard problem unfortunately. Maybe we need to split the HDAs that can be deleted into those that appear as DatasetCollectionElements and those that do not ?

**Review by mvdbeek — COMMENTED** [review]
A quick review here. This looks really good for as far as I can judge the client portion (I'd have used a ConfigProvider instead of the explicit call to `getGalaxyInstance`, but it still looks very clean).

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Ah, and we should also treat HDAs that appear in published histories differently (probably in the listing step, not here. If you send the item id here we should just assume you know what you're doing).

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: That's a great idea. I will do that thanks!

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: I agree these backend changes are a bit simplistic just to get something to test with. I will open separate PRs for them and include your comments. Then we can iterate over those PRs and drop the API changes from this one once they are merged.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: I don't particularly mind having the backend changes in here, if it's not a huge amount of refactoring it helps to see why it was changed.

File: lib/galaxy_test/api/test_datasets.py
Comment (jdavcs) [inline lib/galaxy_test/api/test_datasets.py]:
```suggestion
        for deleted_id in expected_deleted_ids:
            dataset = self._get(f"histories/{history_id}/contents/{deleted_id}").json()
```
Assuming that was the intention?

Comment (davelopez) [reply lib/galaxy_test/api/test_datasets.py]: It was! thanks! :)

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (jdavcs) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Brainstorming.. Unless this has been implemented already (looking at the model, I don't see it), we could add a field to the HDA model that stores a count of its parent collections. Whenever an HDA instance is added to a collection, we increment that field. Deleting a collection decrements the counters of its members. Deleting an HDA is only allowed when the field is 0. Like ref counts in garbage collection. 
(EDIT: i was typing this reply to the very first comment in this thread, without seeing the entire conversation :smile: )

**Comment by natefoo** [comment]
For the record, `allow_user_dataset_purge` exists as an option (and defaulted to `false`) mainly as a safety switch for when the functionality for users purging their own data immediately was added, in case we had some kind of critical bug in that feature. I agree that at this point it can be removed (or at least undocumented and we should not worry about what to do if it is set to `false`).

**Comment by davelopez** [comment]
> I think "old" and "large" are good candidates for user-level preferences: both could mean very different things (quantities) to different users.

This is a good point, I agree these probably should be user preferences with some reasonable defaults.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: That would be a field that potentially sees a lot of concurrent updates for a usecase that is probably not run that often (i suppose you're not constantly hitting the limit / worried about your storage after all).  it's not that hard to generate a query for this. If  a query turns out to be a problem we can still think about alternatives.

Comment (jdavcs) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Oh yes, you're right. And now that I think of it, a query shouldn't be expensive.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Hopefully this is addressed in https://github.com/galaxyproject/galaxy/pull/13113/commits/3502fa15486a2dfbc6bda06864b2173fc9afadff

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: I guess the question is would we want to batch-delete mixed content ? If yes, we should accept a payload like `[{'src': 'hda'}, 'id': 1], {'src': 'ldda', 'id': 2}]` etc. It is less efficient in terms of having to repeat `src`, but on the other hand it does feel consistent with regards to other endpoints ... not sure.

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Ohh I see, then it's better to make it like the other endpoints. Otherwise, it may be too tailored for this particular use case and less flexible.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: (Just to clarify after reading again what I wrote: I should have written "other, single value payloads in other endpoints", not "other endpoints", we don't have many other batch endpoints)

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/datasets.py]: A general remark here: Most of the time (but not all, unfortunately) when we write `dataset_id` we are referring to the `id` column of the `dataset` table. "Datasets" as they appear in histories or libraries are always `{History,LibraryDataset}DatasetAssociation`s, which have a `dataset_id` column that points to the id of the Dataset instance, and in some API endpoints we expose this dataset_id. So if we say `dataset_id` here it isn't clear what this refers to.

My suggestion would be to call this field `id` and to add `src` here as well. The client can then translate this into a a more user-friendly error using these two fields.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: So, to recap and see if I understood correctly, this `delete_batch` datasets endpoint **should assume the caller knows what she/he is doing**. And is **the job of the `fetchItems` query to filter out unwanted datasets**, like "public" or those "part of a collection".
We can always add specific cleanup operations extensions targeting collections or entire histories.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: There are pros and cons to doing it this way.
The downside is that in between fetchItems and calling `delete_batch` the user may have changed something (published a history, added a HDA to a collection) and now their deleting things that should have probably not been deleted ... but that doesn't seem like a very likely scenario.

The pro is that it is simpler and a good enough start IMO.
One way to give stronger guarantees about not deleting unwanted items is to build an atomic update query where you basically use the same query that you'd use to get a list of items that is not in a published history / part of a collection and you turn that into the update statement. That would probably be a significant amount of additional effort.

My gut feeling is that as long as we document in the delete_batch endpoint that no additional checks (beyond permission) are performed that this is good enough, and we could add a "strict" mode later if it becomes necessary.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: (but if you want to implement the "strict" mode or another way to do something like this now that's cool too, it is your call)

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: Thanks! I care a lot about correct naming, but it's usually the hardest part :laughing: 
I changed some of the namings in https://github.com/galaxyproject/galaxy/pull/13113/commits/508926f034a6873cf2dc5836938c18ea654e5bbf, but not sure if they are more clarifying. This one in particular I changed it to `dataset` because now is not just the `id` but also the `src` and is somewhat like a "dataset reference".

Happy to change them further now that we are at it.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/datasets.py]:
```suggestion
            "A list of dataset IDs and the corresponding error message if something "
```
Just to reserve `collection` for dataset collections, even if that is clear here.

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/datasets.py]: That looks good, thank you!

**Comment by davelopez** [comment]
@jmchilton I've just pushed some changes to support multiple quota sources in the dashboard when they are available. This should make https://github.com/galaxyproject/galaxy/pull/10977 hopefully really easy to integrate. Probably just a matter of adapting the `QuotaUsageProvider`.

The current result using some mock data is like this:

Screenshot from 2022-01-07 18-22-48

**Comment by bgruening** [comment]
@jennaj raised the idea to show a panel for hidden datasets/dataset collections over in the goats chat. Not sure if this should/could be added to this PR. But we should keep track of that idea.

**Comment by gallardoalba** [comment]
Great! Also I think that it could be nice if it would be possible to integrate compression tools (e.g. FASTA to FASTA.gz), in order to allow the users to identify which files could be compressed in order to save storage.

**Comment by bgruening** [comment]
> Great! Also I think that it could be nice if it would be possible to integrate compression tools (e.g. FASTA to FASTA.gz), in order to allow the users to identify which files could be compressed in order to save storage.

What do you mean here? That we should have an area where we can give general recommendations? The current dashboard is to free-space and remove files from users-quota. How should we integrate tools here?

Google is showing this:

grafik

We could maybe add a [Galaxy webhook](https://github.com/galaxyproject/galaxy/pull/3040) to this page, so that every Galaxy instance can add dynamic content to this page. E.g. we could add:

* link to a GTN research data management tutorial
* link to a page with general tips to save storage - like your idea of only uploading .gz compressed reads
* link to a quota-increase-form

**Comment by mvdbeek** [comment]
Please keep it simple for a first pass, these are good ideas but this is modular enough to build on, I would suggest opening issues to scope this out further.

**Comment by gallardoalba** [comment]
> What do you mean here? That we should have an area where we can give general recommendations? The current dashboard is to free-space and remove files from users-quota. How should we integrate tools here?

Yes, I know; but compressing file also allows to save space. I just consider that it can be useful since people from VGP also suggested being able to compress their FASTA files in Galaxy.

**Comment by davelopez** [comment]
I think this may be ready for the final review as It is already bigger than I would have desired :sweat_smile: 

I've put some *Beta* indicators in the Storage Manager because I share @mvdbeek's concerns about selecting the appropriate datasets to be deleted but this is probably a good base to start playing around on dev and to build on.

I will be away for a week or so, but I'm happy to fix and improve whatever pop-ups in the review when I'm back.
Thank you all for the comments and help! :beers:

**Comment by davelopez** [comment]
Added a new `Refresh` button that will trigger a recalculation of the user's disk usage.

recalculate

Since right now we cannot know when the recalculation will be ready, the button gets disabled for about 30 seconds and a message will indicate that the user should check back later. This will slightly mitigate accidental multiple recalculations since this operation may be expensive.

I also removed this recalculation from the logout process but I can keep it there if it is somehow important.

**Comment by davelopez** [comment]
> Didn't we make disk recalculation fast? We replaced that query a while back with one that's quite efficient, or am I misremembering?

It may very well be fast and I may have been fooled by the old code comments :laughing: but I don't have enough test data to verify :sweat_smile:

**Comment by hexylena** [comment]
https://github.com/galaxyproject/galaxy/pull/8452 yeah it does bjoern's in 180 ms, and our 'worst' user in 0.5s

**Comment by davelopez** [comment]
> https://github.com/galaxyproject/galaxy/pull/8452 yeah it does bjoern's in 180 ms, and our 'worst' user in 0.5s.

So cool! 

@mvdbeek, is this still something that may be run as an external celery task? or can we just invoke the `recalculate_user_disk_usage` like any other API endpoint so we can get the result right away?

**Comment by mvdbeek** [comment]
I'd prefer this to be async, 0.2-0.5 seconds is still pretty long if someone hammers the endpoint. Not that there aren't easier ways to cause a lot of load, but this is a nice playground for our async task system.

File: client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js
Comment (mvdbeek) [inline client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js:14]: The user quota has a modification time. You could use that to check if the update is complete. But maybe that's premature optimization. Not needed for a first pass.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/controllers/user.py:255]: You'll want to update https://github.com/galaxyproject/galaxy/blob/dev/client/src/mvc/user/user-quotameter.js#L149 as well

File: client/src/mvc/user/user-quotameter.js
Comment (mvdbeek) [inline client/src/mvc/user/user-quotameter.js]: I think the dashboard should always be reachable when you click on the quota

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (mvdbeek) [inline client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue]: We should add this, without sorting this will not work very well.

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (mvdbeek) [inline client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue]: That is too small to be usable IMO
```suggestion
            perPage: 50,
```

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (mvdbeek) [inline client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue:10]: total items maxes out at 500 because of the API limit, there may be more items actually. Not sure how to convery this, but if we sort by size it's maybe not that important

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (mvdbeek) [reply client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue]: 4f6ff859f4ca7a49740a08f3a70219a78c376b9c works for me, but we should add some test there

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/datasets.py:551]: The purge especially could take a while. I think this should ideally be an asynchronous celery endpoint (in a follow up).

**Review by mvdbeek — COMMENTED** [review]
This is working well for me, I'm looking forward to the remaining categories :).

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (davelopez) [reply client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue]: Ups... this is a leftover from testing the pagination, thanks for the catch!

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (davelopez) [reply client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue]: Thank you so much for the fix! I will try to add some tests too :+1:

File: lib/galaxy/webapps/galaxy/services/datasets.py
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/datasets.py:551]: :+1: added to my TODO list for follow-ups

File: client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue
Comment (davelopez) [reply client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue:10]: Oh, you are right, I didn't try with so many datasets. I guess we can set the `limit` query parameter in the requests to a pretty higher number like 10000 to bypass the 500 default limit but still there will be a limit.

The sorting by size seems to be working pretty well though :)

File: client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js
Comment (davelopez) [reply client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js:14]: Ideally, we could get the updated quota value notified from a Vuex store right? Also after the celery task for quota recalculation finishes, I need to get more familiar with those stores yet :sweat_smile: 
So yes, it looks like a good follow up optimization :)

**Comment by davelopez** [comment]
Thank you @mvdbeek for the re-review and help!
I think I've addressed most of the things that you mentioned, let me know if there is something else I can improve in this first iteration.
I'll take it out of draft now since it looks Github Actions are back from the dead again :laughing:

File: client/src/components/User/DiskUsage/Management/services.js
Comment (mvdbeek) [inline client/src/components/User/DiskUsage/Management/services.js]: I think that can be very heavy on the server. I'd either leave this unspecified and adjust the message to say "more than 500 items" or I'd limit this at a lower number and return the total matches in the header.

**Review by mvdbeek — APPROVED** [review]

File: client/src/components/User/DiskUsage/Management/services.js
Comment (davelopez) [reply client/src/components/User/DiskUsage/Management/services.js]: Understand!, I will limit it back to 500 and indicate this limit in the operation description so the user doesn't get confused if more items appear after the first pass. Since it's ordered by size by default it should be good enough to review/bulk delete in chunks of 500.

Comment (davelopez) [reply client/src/components/User/DiskUsage/Management/services.js]: Ok, the limit is now back to 500 and the review dialog will display `500+` and a clarification message in the footer when there are at least 500 items

Screenshot from 2022-03-25 13-15-15

**Comment by mvdbeek** [comment]
Brilliant, this is so cool! Thanks @davelopez!

**Comment by davelopez** [comment]
Thank you so much for all the help and reviews!

**Pull request merged**
